### ISSUE TITLE
Incorrect layer stack hash

### ISSUE THREAD

**Issue by Sopel97** [glinscott/nnue-pytorch#65]
The FC_HASH is hardcoded. It needs to be dynamically computed based on the model layers.

**Comment by vondele** [comment]
hmm, I'd be curious to try some different sizes, any chance for a fix?

**Comment by glinscott** [comment]
Yes, I will take a look.  Until it's fixed, if you want to, you can compile nodchip with the different arch, and print out the hash to workaround the problem - that's how I got the original hash :).

**Comment by vondele** [comment]
I've started an optimization, will serialize later (couldn't figure out where to print it, honestly).

**Comment by Sopel97** [comment]
Layer stack or feature transformer hash: https://github.com/official-stockfish/Stockfish/blob/76daa88cf878b12a03755dc0550b3fa8e4d19cb1/src/nnue/evaluate_nnue.cpp#L67
Combined hash: https://github.com/official-stockfish/Stockfish/blob/76daa88cf878b12a03755dc0550b3fa8e4d19cb1/src/nnue/evaluate_nnue.cpp#L100

**Comment by glinscott** [comment]
Fixed in 401d25b7652a2a5c17897618225bbabd81ac8f03

**Comment by glinscott** [comment]
Note that you will need to add a layer into the layers list in serialize.py, but otherwise it shouldn't need any changes.

**Comment by vondele** [comment]
Thanks! For changes sizes nothing will be needed right?

**Comment by glinscott** [comment]
No changes required if just adjusting size :).

**Comment by vondele** [comment]
I'm not getting it to work, must make a trivial mistake somewhere. I have this change to the trainer:
```diff
diff --git a/model.py b/model.py
index fb34284..c1d4724 100644
--- a/model.py
+++ b/model.py
@@ -6,9 +6,9 @@ import torch.nn.functional as F
 import pytorch_lightning as pl
 
 # 3 layer fully connected network
-L1 = 256
-L2 = 32
-L3 = 32
+L1 = 512
+L2 = 16
+L3 = 16
 
 class NNUE(pl.LightningModule):
   """
```
and this to the player:

```diff
diff --git a/src/evaluate.h b/src/evaluate.h
index 8beca2d0a..92cf96df0 100644
--- a/src/evaluate.h
+++ b/src/evaluate.h
@@ -36,7 +36,7 @@ namespace Eval {
   // The default net name MUST follow the format nn-[SHA256 first 12 digits].nnue
   // for the build process (profile-build and fishtest) to work. Do not change the
   // name of the macro, as it is used in the Makefile.
-  #define EvalFileDefaultName   "nn-62ef826d1a6d.nnue"
+  #define EvalFileDefaultName   "nn-f73eccf09644.nnue"
 
   namespace NNUE {
 
diff --git a/src/nnue/architectures/halfkp_256x2-32-32.h b/src/nnue/architectures/halfkp_256x2-32-32.h
index a0fe2e0ad..410465139 100644
--- a/src/nnue/architectures/halfkp_256x2-32-32.h
+++ b/src/nnue/architectures/halfkp_256x2-32-32.h
@@ -35,14 +35,14 @@ using RawFeatures = Features::FeatureSet<
     Features::HalfKP<Features::Side::kFriend>>;
 
 // Number of input feature dimensions after conversion
-constexpr IndexType kTransformedFeatureDimensions = 256;
+constexpr IndexType kTransformedFeatureDimensions = 512;
 
 namespace Layers {
 
 // Define network structure
 using InputLayer = InputSlice<kTransformedFeatureDimensions * 2>;
-using HiddenLayer1 = ClippedReLU<AffineTransform<InputLayer, 32>>;
-using HiddenLayer2 = ClippedReLU<AffineTransform<HiddenLayer1, 32>>;
+using HiddenLayer1 = ClippedReLU<AffineTransform<InputLayer, 16>>;
+using HiddenLayer2 = ClippedReLU<AffineTransform<HiddenLayer1, 16>>;
 using OutputLayer = AffineTransform<HiddenLayer2, 1>;
 
 }  // namespace Layers
```
but the network fails to load.

A sample checkpoint would be https://drive.google.com/file/d/1iTnABec1Ok771Jo1_kyVEvbvdTYowhtk/view?usp=sharing

Any idea what I might be doing wrong?

**Comment by Sopel97** [comment]
this could be because we're not padding the input dimensions to a multiple of 32 on serialization.

For example if we had a 4->4 affine transform it would naively get serialized like
```
01 02 03 04 
05 06 07 08
09 10 11 12
13 14 15 16
```

but what needs to be done is
```
01 02 03 04 00 00 00 00 ... to 32 bytes
05 06 07 08 00 00 00 00 ... to 32 bytes
09 10 11 12 00 00 00 00 ... to 32 bytes
13 14 15 16 00 00 00 00 ... to 32 bytes
```

**Comment by vondele** [comment]
good analysis! That is likely going to be the reason.

**Comment by glinscott** [comment]
Ah, we didn't need that with the existing layers.  Should be fixed in 36de78f8729fe48f97fe81756ce65e2e5700e4c6

**Comment by vondele** [comment]
@glinscott works well now... thanks!
